File: examples/local-agent/typescript/refund-outcomes.mjs
/** Explicit outcome handling and durable recovery with the Auths local agent. */

import { connect } from "@auths-dev/sdk";
import { Stripe } from "@auths-dev/profile-stripe";

/**
 * Secret store that durably copies and atomically replaces encoded handles
 * before `save` resolves, and deletes them only after a terminal outcome.
 * @typedef {{ save(operationId: string, encoded: ReturnType<import("@auths-dev/sdk").RecoveryHandle["toBytes"]>): Promise<void>, delete(operationId: string): Promise<void> }} RecoveryStore
 */

/**
 * Create one refund. Keep `refundRequestId` stable only across retries of that
 * one intended refund, and supply a deployment secret store for recovery data.
 *
 * @param {RecoveryStore} recoveryStore
 * @param {string} refundRequestId
 */
export async function createRefund(recoveryStore, refundRequestId) {
  const session = await connect();
  try {
    const refunds = new Stripe(session, { connection: "billing" }).refunds;
    let outcome = await refunds.createOutcome(
      {
        paymentIntent: "pi_123",
        amount: 2_000,
        currency: "usd",
      },
      { idempotencyKey: refundRequestId },
    );

    let pendingRecoveryOperation = null;
    if (outcome.kind === "recovery-required") {
      await saveRecovery(recoveryStore, outcome.operationId, outcome.recovery);
      pendingRecoveryOperation = outcome.operationId;
      outcome = await refunds.recoverOutcome(outcome.recovery);
    }

    switch (outcome.kind) {
      case "completed":
        if (pendingRecoveryOperation !== null) {
          await recoveryStore.delete(pendingRecoveryOperation);
        }
        console.log("refund", outcome.value.id);
        console.log(
          "operation",
          outcome.value.auths.operationId,
          outcome.value.auths.completion,
        );
        console.log("receipts", ...outcome.value.auths.receiptIds);
        return;
      case "recovery-required":
        await saveRecovery(recoveryStore, outcome.operationId, outcome.recovery);
        console.log("recovery queued", outcome.operationId, ...outcome.receiptIds);
        return;
      case "receipt-integrity-failed":
        console.error(
          "receipt integrity failure; contact the Auths operator/support",
          outcome.operationId,
          outcome.state,
          outcome.effect,
          outcome.terminal,
          outcome.issue.code,
        );
        return;
      case "conflict":
        await saveRecovery(recoveryStore, outcome.operationId, outcome.recovery);
        throw new Error(
          `refund conflict for original operation ${outcome.operationId}: ${outcome.issue.code}`,
        );
      case "denied":
        if (pendingRecoveryOperation !== null) {
          await recoveryStore.delete(pendingRecoveryOperation);
        }
        throw new Error(
          `refund denied for operation ${outcome.operationId}: ${outcome.issue.code}`,
        );
      case "unavailable":
        throw new Error(
          `refund unavailable before effect (operation=${outcome.operationId}): ${outcome.issue.code}`,
        );
      case "not-applied":
        if (pendingRecoveryOperation !== null) {
          await recoveryStore.delete(pendingRecoveryOperation);
        }
        throw new Error(
          `refund proven not applied for operation ${outcome.operationId} (${outcome.completion}): ${outcome.issue.code}`,
        );
      case "partial":
        if (pendingRecoveryOperation !== null) {
          await recoveryStore.delete(pendingRecoveryOperation);
        }
        throw new Error(
          `refund partially applied for operation ${outcome.operationId} (${outcome.completion}): ${outcome.issue.code}`,
        );
      default:
        throw new Error("unrecognized Auths outcome");
    }
  } finally {
    await session.close();
  }
}

async function saveRecovery(store, operationId, recovery) {
  const encoded = recovery.toBytes();
  try {
    await store.save(operationId, encoded);
  } finally {
    encoded.fill(0);
  }
}

// The application bootstrap calls createRefund with its deployment secret store.

File: examples/local-agent/python/refund_outcomes.py
"""Explicit outcome handling and durable recovery with the Auths local agent."""

from __future__ import annotations

from typing import Optional, Protocol

import auths
from auths.profile_runtime import (
    Completed,
    Conflict,
    Denied,
    NotApplied,
    Partial,
    ReceiptIntegrityFailed,
    RecoveryRequired,
    Unavailable,
)
from auths_profiles.stripe import Stripe


class RecoveryStore(Protocol):
    """Store that durably copies and atomically replaces bytes before returning."""

    async def save(self, operation_id: str, encoded: bytearray) -> None: ...

    async def delete(self, operation_id: str) -> None: ...


async def save_recovery(
    store: RecoveryStore,
    operation_id: str,
    recovery: auths.RecoveryHandle,
) -> None:
    encoded = bytearray(recovery.to_bytes())
    try:
        await store.save(operation_id, encoded)
    finally:
        encoded[:] = b"\0" * len(encoded)


async def create_refund(
    recovery_store: RecoveryStore,
    *,
    refund_request_id: str,
) -> None:
    """Create one refund; keep ``refund_request_id`` stable only for its retries."""

    async with auths.connect() as session:
        refunds = Stripe(session, connection="billing").refunds
        outcome = await refunds.create_outcome(
            payment_intent="pi_123",
            amount=2_000,
            currency="usd",
            options=auths.OperationOptions(idempotency_key=refund_request_id),
        )

        pending_recovery: Optional[str] = None
        if isinstance(outcome, RecoveryRequired):
            await save_recovery(
                recovery_store, outcome.operation_id, outcome.recovery
            )
            pending_recovery = outcome.operation_id
            outcome = await refunds.recover_outcome(outcome.recovery)

        if isinstance(outcome, Completed):
            refund = outcome.value
            if pending_recovery is not None:
                await recovery_store.delete(pending_recovery)
            print("refund", refund.id)
            print("operation", refund.auths.operation_id, refund.auths.completion)
            print("receipts", *refund.auths.receipt_ids)
            return

        if isinstance(outcome, RecoveryRequired):
            await save_recovery(
                recovery_store, outcome.operation_id, outcome.recovery
            )
            print("recovery queued", outcome.operation_id, *outcome.receipt_ids)
            return

        if isinstance(outcome, ReceiptIntegrityFailed):
            print(
                "receipt integrity failure; contact the Auths operator/support",
                outcome.operation_id,
                outcome.state,
                outcome.effect,
                outcome.terminal,
                outcome.issue.code,
            )
            return

        if isinstance(outcome, Conflict):
            await save_recovery(
                recovery_store, outcome.operation_id, outcome.recovery
            )
            raise RuntimeError(
                f"refund conflict for original operation {outcome.operation_id}: "
                f"{outcome.issue.code}"
            )

        if isinstance(outcome, Denied):
            if pending_recovery is not None:
                await recovery_store.delete(pending_recovery)
            raise RuntimeError(
                f"refund denied for operation {outcome.operation_id}: {outcome.issue.code}"
            )

        if isinstance(outcome, Unavailable):
            raise RuntimeError(
                f"refund unavailable before effect (operation={outcome.operation_id}): "
                f"{outcome.issue.code}"
            )

        if isinstance(outcome, NotApplied):
            if pending_recovery is not None:
                await recovery_store.delete(pending_recovery)
            raise RuntimeError(
                f"refund proven not applied for operation {outcome.operation_id} "
                f"({outcome.completion}): {outcome.issue.code}"
            )

        if isinstance(outcome, Partial):
            if pending_recovery is not None:
                await recovery_store.delete(pending_recovery)
            raise RuntimeError(
                f"refund partially applied for operation {outcome.operation_id} "
                f"({outcome.completion}): {outcome.issue.code}"
            )

        raise AssertionError("unrecognized Auths outcome")


# The application bootstrap calls create_refund with its deployment secret store.
